### DISCUSSION TITLE
useFieldArray with empty array - no key when submitting form

### DISCUSSION THREAD

**Discussion by idlefingers** [react-hook-form/react-hook-form#1886] [Q&A] [answered]
When using `useFieldArray`, if you submit a form with no items in the field array, the key is entirely omitted from the data object passed to the submit handler function.

For example, in the sandbox from the docs for useFieldArray, https://codesandbox.io/s/react-hook-form-usefieldarray-vy8fv?file=/src/index.js, if you submit with an item in the field array, you get what you'd expect:

```
{
  test: [
    { firstName: 'x', lastName: 'y' }
  ]
}
```

But if you remove that item and submit, the data is just an empty object. I'd expect the `test` key to still exist, but with an empty array as the value:

```
{
  test: []
}
```

Is this a bug, or am I just missing something? I'd rather not introduce a bunch of hacks in my submit handler to make empty arrays if the key is missing, but from what I can gather from the docs, there's no other option?

**Comment by bluebill1049** [comment, marked as answer]
This is not a bug, because `useFieldArray` is uncontrolled (native form), so input state lives in inputs, think as when you building native inputs. when all inputs remove, then you won't have any data.

**Reply by idlefingers** [reply]
Ok, thanks for the clarification. I guess there's no way attaching `useFieldArray` to the control could set the default value to `[]`?

With 'normal' fields, it makes sense that if they're not there, they're not there, but arrays feel a little different to me - I mean, the array is there, but there's just no items in it..

**Reply by bluebill1049** [reply]
i agree with you, perhaps you can have a `useEffect` to check if empty field do a `setValue`? (I never try that just imagine in my head)

**Reply by idlefingers** [reply]
https://codesandbox.io/s/react-hook-form-usefieldarray-53xfp?file=/src/index.js like this? Still no `test: []` in the submitted data :(

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/react-hook-form-usefieldarray-eqwbr?file=/src/index.js

**Reply by bluebill1049** [reply]
not the best, but it works :)

**Reply by jpodpro** [reply]
@idlefingers did you come up with a non-hack solution to this? feels like a pretty obvious problem to me.

**Reply by idlefingers** [reply]
Nope, I went back to hacks in my submit handler. Custom field array handlers were even more brittle. :/

**Reply by prince272** [reply]
Thanks for your solution @bluebill1049.  I was trying stuffs like this which even didn't work. 
`<input name={'test[0]'} type="hidden" ref={form.register()} /> `

**Reply by prince272** [reply]
Wait... It does work with other higher versions of react hook form. From react hook form version  6.9.3 and above.

**Reply by yiy0ung** [reply]
Is there any other solution? The above solution is not easy to use in complex forms.

**Reply by danedavid** [reply]
@bluebill1049 The proposed solution leads to infinite re-renders for v6 & v7:
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-nmxq4
